rules.go

```diff
@@ -0,0 +1,482 @@
+	options     map[string]bool
+	isException bool
+	domains     map[string]bool
+}
+
+func parseRule(ruleText string) (*ruleAdBlock, error) {
+	if strings.TrimSpace(ruleText) == "" {
+		return nil, ErrEmptyLine
+	}
+	rule := &ruleAdBlock{
+		domains:  map[string]bool{},
+		options:  map[string]bool{},
+		raw:      ruleText,
+		ruleText: strings.TrimSpace(ruleText),
+	}
+
+	isComment := strings.Contains(rule.ruleText, "!") || strings.Contains(rule.ruleText, "[Adblock")
+	if isComment {
+		return nil, ErrSkipComment
+	}
+
+	isHTMLRule := strings.Contains(rule.ruleText, "##") || strings.Contains(rule.ruleText, "#@#")
+	if isHTMLRule {
+		return nil, ErrSkipHTML
+	}
+
+	rule.isException = strings.HasPrefix(rule.ruleText, "@@")
+	if rule.isException {
+		rule.ruleText = rule.ruleText[2:]
+	}
+
+	if strings.Contains(rule.ruleText, "$") {
+		parts := strings.SplitN(rule.ruleText, "$", 2)
+		length := len(parts)
+
+		if length > 0 {
+			rule.ruleText = parts[0]
+		}
+
+		if length > 1 {
+			for _, option := range strings.Split(parts[1], ",") {
+				if strings.HasPrefix(option, "domain=") {
+					rule.domains = parseDomainOption(option)
+				} else {
+					optionName := strings.TrimPrefix(option, "~")
+					if ok := strings.Contains(supportedOptionsPat, optionName); !ok {
+						return nil, ErrUnsupportedRule
+					}
+					rule.options[optionName] = !strings.HasPrefix(option, "~")
+				}
+			}
+		}
+	}
+
+	rule.regexString = ruleToRegexp(rule.ruleText)
+
+	return rule, nil
+}
+
+type RuleSet struct {
+	whitelist      []*ruleAdBlock
+	whitelistRegex *regexp.Regexp
+	blacklist      []*ruleAdBlock
+	blacklistRegex *regexp.Regexp
+
+	whitelistDomainsNoOptions    []*ruleAdBlock
+	whitelistDomainsWithOptions  []*ruleAdBlock
+	whitelistIncludeDomains      map[string][]*ruleAdBlock
+	whitelistIncludeDomainsRegex map[string]*regexp.Regexp
+	whitelistExcludeDomains      map[string][]*ruleAdBlock
+	whitelistExcludeDomainsRegex map[string]*regexp.Regexp
+	blacklistDomainsNoOptions    []*ruleAdBlock
+	blacklistDomainsWithOptions  []*ruleAdBlock
+	blacklistIncludeDomains      map[string][]*ruleAdBlock
+	blacklistIncludeDomainsRegex map[string]*regexp.Regexp
+	blacklistExcludeDomains      map[string][]*ruleAdBlock
+	blacklistExcludeDomainsRegex map[string]*regexp.Regexp
+
+	whitelistIncludeOptions      map[string][]*ruleAdBlock
+	whitelistIncludeOptionsRegex map[string]*regexp.Regexp
+	whitelistExcludeOptions      map[string][]*ruleAdBlock
+	whitelistExcludeOptionsRegex map[string]*regexp.Regexp
+	blacklistIncludeOptions      map[string][]*ruleAdBlock
+	blacklistIncludeOptionsRegex map[string]*regexp.Regexp
+	blacklistExcludeOptions      map[string][]*ruleAdBlock
+	blacklistExcludeOptionsRegex map[string]*regexp.Regexp
+}
+
+func matchWhite(ruleSet RuleSet, req Request) bool {
+	if ruleSet.whitelistRegex != nil && ruleSet.whitelistRegex.MatchString(req.URL.String()) {
+		return true
+	}
+
+	options := extractOptionsFromRequest(req)
+	for option, active := range options {
```

Comment:
The logic was:
If the URL ends with `.js`, then we know that it can be matched against rules that expect the link to be a script.
If the URL does not end with `.js`, then we know that it can be matched against rules that expect the link not to be a script.
besides that, it also will be matched against rules that do not specify if it can be a script or not.
Anyway, now the logic is different using a prefix tree.